okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java

```diff
@@ -0,0 +1,144 @@
+ *     <li>
+ *         Each such request gets scheduled according to an exponential backoff algorithm.
+ *     </li>
+ * </ol>
+ * <p>
+ * This class provides an asynchronous implementation of Netflix's
+ * <a href="https://github.com/Netflix/concurrency-limits/">concurrency-limits</a> library for determining the
+ * above mentioned concurrency estimates.
+ * <p>
+ * In order to use this class, one should acquire a Limiter for their request, which returns a future. once the Future
+ * is completed, the caller can assume that the request is schedulable. After the request completes, the caller
+ * <b>must</b> call one of the methods on {@link Limiter.Listener} in order to provide feedback about the request's
+ * success. If this is not done, a deadlock could result.
+ */
+final class ConcurrencyLimiters {
+    private static final String FALLBACK = "";
+
+    private final ConcurrentMap<String, ConcurrencyLimiter> limiters = new ConcurrentHashMap<>();
+
+    @VisibleForTesting
+    ConcurrencyLimiter limiter(String name) {
+        return limiters.computeIfAbsent(name, key -> new ConcurrencyLimiter(DefaultLimiter.newBuilder()
+                .limit(TracingLimitDecorator.wrap(AIMDLimit.newBuilder().initialLimit(1).build()))
+                .build(new SimpleStrategy<>())));
+    }
+
+    ConcurrencyLimiter limiter(Request request) {
+        final String limiterKey;
+        String pathTemplate = request.header(OkhttpTraceInterceptor.PATH_TEMPLATE_HEADER);
+        if (pathTemplate == null) {
+            limiterKey = FALLBACK;
+        } else {
+            limiterKey = request.method() + " " + pathTemplate;
+        }
+        return limiter(limiterKey);
+    }
+
+    /**
+     * The Netflix library provides either a blocking approach or a non-blocking approach which might say
+     * you can't be scheduled at this time. All of our HTTP calls are asynchronous, so we really want to get
+     * a {@link ListenableFuture} that we can add a callback to. This class then is a translation of
+     * {@link com.netflix.concurrency.limits.limiter.BlockingLimiter} to be asynchronous, maintaining a queue
+     * of currently waiting requests.
+     * <p>
+     * Upon a request finishing, we check if there are any waiting requests, and if there are we attempt to trigger
+     * some more.
+     */
+    static final class ConcurrencyLimiter {
+        @GuardedBy("this")
+        private final Queue<SettableFuture<Limiter.Listener>> waitingRequests = new ArrayDeque<>();
+        private final Limiter<Void> limiter;
+
+        @VisibleForTesting
+        ConcurrencyLimiter(Limiter<Void> limiter) {
+            this.limiter = limiter;
+        }
+
+        synchronized ListenableFuture<Limiter.Listener> acquire() {
+            SettableFuture<Limiter.Listener> future = SettableFuture.create();
+            waitingRequests.add(future);
+            processQueue();
+            return future;
+        }
+
+        private synchronized void processQueue() {
```

Comment:
personally had a hard time understanding how this works/when this gets called given it can be called either when you `acquire` or when you notify the listener that you're done.

Reply:
Chatted to joe in person, he's now ok with this I think.